In [2]:
%load_ext sql

In [4]:
%sql mysql+pymysql://root:Mysql.003@localhost:3306/md_water_services

In [5]:
%%sql

SELECT 
    *
FROM employee LIMIT 5;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[(0, 'Amara Jengo', '+99637993287 ', None, '36 Pwani Mchangani Road', 'Sokoto', 'Ilanga', 'Field Surveyor'),
 (1, 'Bello Azibo', '+99643864786 ', None, '129 Ziwa La Kioo Road', 'Kilimani', 'Rural', 'Field Surveyor'),
 (2, 'Bakari Iniko', '+99222599041 ', None, '18 Mlima Tazama Avenue', 'Hawassa', 'Rural', 'Field Surveyor'),
 (3, 'Malachi Mavuso', '+99945849900 ', None, '100 Mogadishu Road', 'Akatsi', 'Lusaka', 'Field Surveyor'),
 (4, 'Cheche Buhle', '+99381679640 ', None, '1 Savanna Street', 'Akatsi', 'Rural', 'Field Surveyor')]

In [6]:
%%sql
SHOW COLUMNS
FROM employee;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
8 rows affected.


[('assigned_employee_id', 'int(11)', 'NO', 'PRI', None, ''),
 ('employee_name', 'varchar(255)', 'YES', '', None, ''),
 ('phone_number', 'varchar(15)', 'YES', '', None, ''),
 ('email', 'varchar(255)', 'YES', '', None, ''),
 ('address', 'varchar(255)', 'YES', '', None, ''),
 ('province_name', 'varchar(255)', 'YES', '', None, ''),
 ('town_name', 'varchar(255)', 'YES', '', None, ''),
 ('position', 'varchar(255)', 'YES', '', None, '')]

In [10]:
%%sql

SELECT 
  employee_name,
  CONCAT(LOWER(REPLACE(employee_name, ' ', '.')), '@ndogowater.gov') AS email
FROM employee
LIMIT 10;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
10 rows affected.


[('Amara Jengo', 'amara.jengo@ndogowater.gov'),
 ('Bello Azibo', 'bello.azibo@ndogowater.gov'),
 ('Bakari Iniko', 'bakari.iniko@ndogowater.gov'),
 ('Malachi Mavuso', 'malachi.mavuso@ndogowater.gov'),
 ('Cheche Buhle', 'cheche.buhle@ndogowater.gov'),
 ('Zuriel Matembo', 'zuriel.matembo@ndogowater.gov'),
 ('Deka Osumare', 'deka.osumare@ndogowater.gov'),
 ('Lalitha Kaburi', 'lalitha.kaburi@ndogowater.gov'),
 ('Enitan Zuri', 'enitan.zuri@ndogowater.gov'),
 ('Farai Nia', 'farai.nia@ndogowater.gov')]

In [11]:
%%sql

UPDATE employee
SET email = CONCAT(LOWER(REPLACE(employee_name, ' ', '.')), '@ndogowater.gov');

 * mysql+pymysql://root:***@localhost:3306/md_water_services
56 rows affected.


[]

In [13]:
%%sql
SELECT * FROM employee LIMIT 5;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[(0, 'Amara Jengo', '+99637993287 ', 'amara.jengo@ndogowater.gov', '36 Pwani Mchangani Road', 'Sokoto', 'Ilanga', 'Field Surveyor'),
 (1, 'Bello Azibo', '+99643864786 ', 'bello.azibo@ndogowater.gov', '129 Ziwa La Kioo Road', 'Kilimani', 'Rural', 'Field Surveyor'),
 (2, 'Bakari Iniko', '+99222599041 ', 'bakari.iniko@ndogowater.gov', '18 Mlima Tazama Avenue', 'Hawassa', 'Rural', 'Field Surveyor'),
 (3, 'Malachi Mavuso', '+99945849900 ', 'malachi.mavuso@ndogowater.gov', '100 Mogadishu Road', 'Akatsi', 'Lusaka', 'Field Surveyor'),
 (4, 'Cheche Buhle', '+99381679640 ', 'cheche.buhle@ndogowater.gov', '1 Savanna Street', 'Akatsi', 'Rural', 'Field Surveyor')]

In [21]:
%%sql

SELECT 
    LENGTH(phone_number)
FROM employee
LIMIT 5;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[(13,), (13,), (13,), (13,), (13,)]

In [24]:
%%sql

SELECT 
    phone_number,
    TRIM(phone_number) AS trim_phone_number,
    LENGTH(TRIM(phone_number)) AS len_phone_number
FROM employee
LIMIT 5;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[('+99637993287 ', '+99637993287', 12),
 ('+99643864786 ', '+99643864786', 12),
 ('+99222599041 ', '+99222599041', 12),
 ('+99945849900 ', '+99945849900', 12),
 ('+99381679640 ', '+99381679640', 12)]

In [25]:
%%sql

UPDATE employee
SET phone_number = TRIM(phone_number);

 * mysql+pymysql://root:***@localhost:3306/md_water_services
56 rows affected.


[]

In [26]:
%%sql
SELECT * FROM employee LIMIT 5;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[(0, 'Amara Jengo', '+99637993287', 'amara.jengo@ndogowater.gov', '36 Pwani Mchangani Road', 'Sokoto', 'Ilanga', 'Field Surveyor'),
 (1, 'Bello Azibo', '+99643864786', 'bello.azibo@ndogowater.gov', '129 Ziwa La Kioo Road', 'Kilimani', 'Rural', 'Field Surveyor'),
 (2, 'Bakari Iniko', '+99222599041', 'bakari.iniko@ndogowater.gov', '18 Mlima Tazama Avenue', 'Hawassa', 'Rural', 'Field Surveyor'),
 (3, 'Malachi Mavuso', '+99945849900', 'malachi.mavuso@ndogowater.gov', '100 Mogadishu Road', 'Akatsi', 'Lusaka', 'Field Surveyor'),
 (4, 'Cheche Buhle', '+99381679640', 'cheche.buhle@ndogowater.gov', '1 Savanna Street', 'Akatsi', 'Rural', 'Field Surveyor')]

In [29]:
%%sql

SELECT 
    DISTINCT town_name,
    COUNT(*) AS employee_count
FROM employee
GROUP BY town_name;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
9 rows affected.


[('Dahabu', 6),
 ('Harare', 5),
 ('Ilanga', 3),
 ('Kintampo', 1),
 ('Lusaka', 4),
 ('Rural', 29),
 ('Serowe', 3),
 ('Yaounde', 1),
 ('Zanzibar', 4)]

In [136]:
%%sql

-- Top 3 Field Surveyors

SELECT 
    e.assigned_employee_id,
    e.employee_name,
    e.phone_number,
    e.email,
    SUM(v.visit_count) AS number_of_visits
FROM employee e
JOIN visits v ON v.assigned_employee_id = e.assigned_employee_id
GROUP BY e.assigned_employee_id
ORDER BY number_of_visits ASC
LIMIT 3;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
3 rows affected.


[(20, 'Kunto Asha', '+99176320477', 'kunto.asha@ndogowater.gov', Decimal('43')),
 (22, 'Lesedi Kofi', '+99611183730', 'lesedi.kofi@ndogowater.gov', Decimal('339')),
 (44, 'Wambui Jabali', '+99269319001', 'wambui.jabali@ndogowater.gov', Decimal('470'))]

In [38]:
%%sql

-- Look up employee info using top 3 IDs with highest number of visits

SELECT 
    assigned_employee_id,
    employee_name,
    email,
    phone_number
FROM employee
WHERE assigned_employee_id IN (1, 30, 34);

 * mysql+pymysql://root:***@localhost:3306/md_water_services
3 rows affected.


[(1, 'Bello Azibo', 'bello.azibo@ndogowater.gov', '+99643864786'),
 (30, 'Pili Zola', 'pili.zola@ndogowater.gov', '+99822478933'),
 (34, 'Rudo Imani', 'rudo.imani@ndogowater.gov', '+99046972648')]

In [140]:
%%sql

SELECT 
    address,
    TRIM(address) AS trimmed_address
FROM employee
WHERE 
    employee_name = "Farai Nia"

 * mysql+pymysql://root:***@localhost:3306/md_water_services
1 rows affected.


[('33 Angélique Kidjo Avenue', '33 Angélique Kidjo Avenue')]

In [142]:
%%sql

SELECT 
    COUNT(*) AS number_of_records
FROM employee
WHERE 
    town_name = "Harare"
    AND province_name = "Kilimani";

 * mysql+pymysql://root:***@localhost:3306/md_water_services
1 rows affected.


[(2,)]

In [43]:
%%sql

-- Number of records per province

SELECT
    province_name,
    COUNT(*) AS number_of_records
FROM employee
GROUP BY province_name
ORDER BY number_of_records ASC;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[('Amanzi', 7),
 ('Sokoto', 9),
 ('Kilimani', 12),
 ('Akatsi', 13),
 ('Hawassa', 15)]

In [143]:
%%sql

SELECT
    SUM(number_of_people_served) AS population_served
FROM
    water_source
WHERE
    type_of_water_source LIKE "%tap%"
ORDER BY
    population_served;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
1 rows affected.


[(Decimal('20423872'),)]

In [44]:
%%sql

SELECT 
    province_name,
    town_name,
    COUNT(*) AS records_per_town
FROM employee
GROUP BY province_name, town_name
ORDER BY province_name ASC, records_per_town DESC;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
15 rows affected.


[('Akatsi', 'Rural', 5),
 ('Akatsi', 'Lusaka', 4),
 ('Akatsi', 'Harare', 3),
 ('Akatsi', 'Kintampo', 1),
 ('Amanzi', 'Dahabu', 6),
 ('Amanzi', 'Rural', 1),
 ('Hawassa', 'Rural', 7),
 ('Hawassa', 'Zanzibar', 4),
 ('Hawassa', 'Serowe', 3),
 ('Hawassa', 'Yaounde', 1),
 ('Kilimani', 'Rural', 9),
 ('Kilimani', 'Harare', 2),
 ('Kilimani', 'Ilanga', 1),
 ('Sokoto', 'Rural', 7),
 ('Sokoto', 'Ilanga', 2)]

In [50]:
%%sql

DESCRIBE location;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[('location_id', 'varchar(255)', 'NO', 'PRI', None, ''),
 ('address', 'varchar(255)', 'YES', '', None, ''),
 ('province_name', 'varchar(255)', 'YES', '', None, ''),
 ('town_name', 'varchar(255)', 'YES', '', None, ''),
 ('location_type', 'varchar(255)', 'YES', '', None, '')]

In [51]:
%%sql

DROP TABLE IF EXISTS location_props;
CREATE TABLE location_props (
    id INT AUTO_INCREMENT PRIMARY KEY,
    location_type VARCHAR(100),
    number_of_records INT
);

INSERT INTO location_props (location_type, number_of_records)
SELECT
    location_type,
    COUNT(*) AS number_of_records
FROM location
GROUP BY location_type;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
0 rows affected.
0 rows affected.
2 rows affected.


[]

In [54]:
%%sql

SELECT 
    location_type,
    number_of_records,
    number_of_records / SUM(number_of_records) AS proportions
FROM location_props;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
1 rows affected.


[('Rural', 23740, Decimal('0.5987'))]

In [57]:
%%sql

-- Proportions of Location Type

SELECT 
    location_type,
    number_of_records,
    number_of_records / total.total_records AS proportions
FROM location_props,
     (SELECT SUM(number_of_records) AS total_records FROM location_props) AS total;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
2 rows affected.


[('Rural', 23740, Decimal('0.5987')), ('Urban', 15910, Decimal('0.4013'))]

In [58]:
%%sql

SELECT SUM(number_of_records) AS total_records FROM location_props;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
1 rows affected.


[(Decimal('39650'),)]

In [59]:
%%sql
DESCRIBE water_source;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
3 rows affected.


[('source_id', 'varchar(510)', 'NO', 'PRI', None, ''),
 ('type_of_water_source', 'varchar(255)', 'YES', '', None, ''),
 ('number_of_people_served', 'int(11)', 'YES', '', None, '')]

In [60]:
%%sql
SELECT * FROM water_source LIMIT 5;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[('AkHa00000224', 'tap_in_home', 956),
 ('AkHa00001224', 'tap_in_home_broken', 930),
 ('AkHa00002224', 'tap_in_home_broken', 486),
 ('AkHa00003224', 'well', 364),
 ('AkHa00004224', 'tap_in_home_broken', 942)]

In [62]:
%%sql

SELECT SUM(number_of_people_served) AS total_number_of_people_served
FROM water_source;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
1 rows affected.


[(Decimal('27628140'),)]

In [68]:
%%sql

-- Counts the number of records per type of water source
-- Calculates the total number of people served per to each type of water source

SELECT 
    type_of_water_source,
    COUNT(*) AS number_of_records,
    SUM(number_of_people_served) AS number_of_people_served
FROM water_source
GROUP BY type_of_water_source
ORDER BY number_of_records;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[('river', 3379, Decimal('2362544')),
 ('shared_tap', 5767, Decimal('11945272')),
 ('tap_in_home_broken', 5856, Decimal('3799720')),
 ('tap_in_home', 7265, Decimal('4678880')),
 ('well', 17383, Decimal('4841724'))]

In [65]:
%%sql

-- Average number of people served by each water source

SELECT 
    type_of_water_source,
    COUNT(*) AS number_of_records,
    ROUND(AVG(number_of_people_served)) AS avg_number_of_people_served
FROM water_source
GROUP BY type_of_water_source
ORDER BY avg_number_of_people_served;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[('well', 17383, Decimal('279')),
 ('tap_in_home', 7265, Decimal('644')),
 ('tap_in_home_broken', 5856, Decimal('649')),
 ('river', 3379, Decimal('699')),
 ('shared_tap', 5767, Decimal('2071'))]

In [67]:
%%sql

SELECT 
    COUNT(*) AS well_count,
    SUM(number_of_people_served) AS number_of_records
FROM water_source
WHERE type_of_water_source = 'well';

 * mysql+pymysql://root:***@localhost:3306/md_water_services
1 rows affected.


[(17383, Decimal('4841724'))]

These results are telling us that 644 people share a tap_in_home on average. Does that make sense?

No it doesn’t, right?
Remember I told you a few important things that apply to tap_in_home and broken_tap_in_home? The surveyors combined the data of many households together and added this as a single tap record, but each household actually has its own tap. In addition to this, there is an average of 6 people living in a home. So 6 people actually share 1 tap (not 644).

In [70]:
%%sql

-- Calculates the total number of people served per to each type of water source

SELECT 
    type_of_water_source,
    SUM(number_of_people_served) AS number_of_people_served
FROM water_source
GROUP BY type_of_water_source
ORDER BY number_of_people_served;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[('river', Decimal('2362544')),
 ('tap_in_home_broken', Decimal('3799720')),
 ('tap_in_home', Decimal('4678880')),
 ('well', Decimal('4841724')),
 ('shared_tap', Decimal('11945272'))]

In [93]:
%%sql

SELECT 
    type_of_water_source,
    SUM(number_of_people_served) AS number_of_people_served,
    
    -- Proportions with up to 4 decimals, multiplied before rounding
    ROUND((COUNT(*) / total.total_sources) * 100, 1) AS proportions_percent,
    
    -- Rounded to nearest whole percent
    ROUND((COUNT(*) / total.total_sources) * 100, 0) AS proportions_rounded
FROM
    water_source,
    (SELECT COUNT(*) AS total_sources FROM water_source) AS total
GROUP BY type_of_water_source;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[('river', Decimal('2362544'), Decimal('8.5'), Decimal('9')),
 ('shared_tap', Decimal('11945272'), Decimal('14.5'), Decimal('15')),
 ('tap_in_home', Decimal('4678880'), Decimal('18.3'), Decimal('18')),
 ('tap_in_home_broken', Decimal('3799720'), Decimal('14.8'), Decimal('15')),
 ('well', Decimal('4841724'), Decimal('43.8'), Decimal('44'))]

In [97]:
%%sql 

SELECT
    type_of_water_source,
    SUM(number_of_people_served) AS total_people_served,
    RANK() OVER (
        ORDER BY SUM(number_of_people_served) DESC
    ) AS usage_rank
FROM water_source
WHERE type_of_water_source <> "tap_in_home"
GROUP BY type_of_water_source
ORDER BY usage_rank;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
4 rows affected.


[('shared_tap', Decimal('11945272'), 1),
 ('well', Decimal('4841724'), 2),
 ('tap_in_home_broken', Decimal('3799720'), 3),
 ('river', Decimal('2362544'), 4)]

In [100]:
%%sql

SELECT 
    type_of_water_source,
    total_people_served,
    DENSE_RANK() OVER (
        ORDER BY total_people_served DESC
    ) AS usage_rank
FROM (
    SELECT 
        type_of_water_source,
        SUM(number_of_people_served) AS total_people_served
    FROM water_source
    WHERE type_of_water_source <> 'tap_in_home'
    GROUP BY type_of_water_source
) AS ranked_sources
ORDER BY usage_rank;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
4 rows affected.


[('shared_tap', Decimal('11945272'), 1),
 ('well', Decimal('4841724'), 2),
 ('tap_in_home_broken', Decimal('3799720'), 3),
 ('river', Decimal('2362544'), 4)]

In [107]:
%%sql

SELECT 
    source_id,
    type_of_water_source,
    number_of_people_served,
    DENSE_RANK() OVER (
        ORDER BY SUM(number_of_people_served) OVER (PARTITION BY source_id) DESC
    ) AS priority_rank
FROM water_source
WHERE type_of_water_source <> 'tap_in_home'
ORDER BY priority_rank;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
(pymysql.err.OperationalError) (4015, 'Window function is allowed only in SELECT list and ORDER BY clause')
[SQL: SELECT 
    source_id,
    type_of_water_source,
    number_of_people_served,
    DENSE_RANK() OVER (
        ORDER BY SUM(number_of_people_served) OVER (PARTITION BY source_id) DESC
    ) AS priority_rank
FROM water_source
WHERE type_of_water_source <> 'tap_in_home'
ORDER BY priority_rank;]
(Background on this error at: https://sqlalche.me/e/20/e3q8)


In [112]:
%%sql

SELECT 
    source_id,
    type_of_water_source,
    total_people_served,
    DENSE_RANK() OVER (
        ORDER BY total_people_served DESC
    ) AS priority_rank
FROM (
    SELECT 
        source_id,
        type_of_water_source,
        SUM(number_of_people_served) AS total_people_served
    FROM water_source
    WHERE type_of_water_source <> 'tap_in_home'
    GROUP BY source_id, type_of_water_source
) AS source_summary
ORDER BY priority_rank ASC, source_id ASC, type_of_water_source ASC
LIMIT 10;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
10 rows affected.


[('AkRu05603224', 'shared_tap', Decimal('3998'), 1),
 ('HaRu19509224', 'shared_tap', Decimal('3998'), 1),
 ('AkRu04862224', 'shared_tap', Decimal('3996'), 2),
 ('AmAs10911224', 'shared_tap', Decimal('3996'), 2),
 ('KiHa22867224', 'shared_tap', Decimal('3996'), 2),
 ('HaRu19839224', 'shared_tap', Decimal('3994'), 3),
 ('KiZu31330224', 'shared_tap', Decimal('3994'), 3),
 ('KiRu28630224', 'shared_tap', Decimal('3992'), 4),
 ('KiZu31415224', 'shared_tap', Decimal('3992'), 4),
 ('AmPw12653224', 'shared_tap', Decimal('3990'), 5)]

In [106]:
%%sql

SELECT 
    type_of_water_source,
    total_people_served,
    ROW_NUMBER() OVER (
        ORDER BY total_people_served DESC
    ) AS usage_rank
FROM (
    SELECT 
        type_of_water_source,
        SUM(number_of_people_served) AS total_people_served
    FROM water_source
    WHERE type_of_water_source <> 'tap_in_home'
    GROUP BY type_of_water_source
) AS ranked_sources
ORDER BY usage_rank;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
4 rows affected.


[('shared_tap', Decimal('11945272'), 1),
 ('well', Decimal('4841724'), 2),
 ('tap_in_home_broken', Decimal('3799720'), 3),
 ('river', Decimal('2362544'), 4)]

In [114]:
%%sql

SELECT 
    source_id,
    type_of_water_source,
    total_people_served,
    ROW_NUMBER() OVER (
        ORDER BY total_people_served DESC
    ) AS priority_rank
FROM (
    SELECT 
        source_id,
        type_of_water_source,
        SUM(number_of_people_served) AS total_people_served
    FROM water_source
    WHERE type_of_water_source <> 'tap_in_home'
    GROUP BY source_id, type_of_water_source
) AS source_summary
ORDER BY priority_rank ASC
LIMIT 10;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
10 rows affected.


[('AkRu05603224', 'shared_tap', Decimal('3998'), 1),
 ('HaRu19509224', 'shared_tap', Decimal('3998'), 2),
 ('AkRu04862224', 'shared_tap', Decimal('3996'), 3),
 ('AmAs10911224', 'shared_tap', Decimal('3996'), 4),
 ('KiHa22867224', 'shared_tap', Decimal('3996'), 5),
 ('HaRu19839224', 'shared_tap', Decimal('3994'), 6),
 ('KiZu31330224', 'shared_tap', Decimal('3994'), 7),
 ('KiRu28630224', 'shared_tap', Decimal('3992'), 8),
 ('KiZu31415224', 'shared_tap', Decimal('3992'), 9),
 ('AmPw12653224', 'shared_tap', Decimal('3990'), 10)]

In [115]:
%%sql

DESCRIBE visits;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
7 rows affected.


[('record_id', 'int(11)', 'NO', 'PRI', None, ''),
 ('location_id', 'varchar(255)', 'YES', 'MUL', None, ''),
 ('source_id', 'varchar(510)', 'YES', 'MUL', None, ''),
 ('time_of_record', 'datetime', 'YES', '', None, ''),
 ('visit_count', 'int(11)', 'YES', '', None, ''),
 ('time_in_queue', 'int(11)', 'YES', '', None, ''),
 ('assigned_employee_id', 'int(11)', 'YES', 'MUL', None, '')]

In [116]:
%%sql

-- Duration of the survey in days

SELECT 
    MIN(time_of_record) AS start_date,
    MAX(time_of_record) AS end_date,
    DATEDIFF(MAX(time_of_record), MIN(time_of_record)) AS survey_duration_days
FROM visits;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
1 rows affected.


[(datetime.datetime(2021, 1, 1, 9, 10), datetime.datetime(2023, 7, 14, 13, 53), 924)]

In [117]:
%%sql

-- Average total queue time for water

SELECT 
    ROUND(AVG(NULLIF(time_in_queue, 0)), 2) AS avg_queue_time_minutes
FROM visits;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
1 rows affected.


[(Decimal('123.26'),)]

In [124]:
%%sql

-- Average queue time on different days, Saturday with highest time 246

SELECT
    time_in_queue,
    time_of_record,
    ROUND(AVG(NULLIF(time_in_queue, 0)), 0) AS avg_queue_time_in_days,
    DAYNAME(time_of_record) AS day_of_week
FROM visits
GROUP BY day_of_week
ORDER BY
    FIELD(DAYNAME(time_of_record), 
        'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday');

 * mysql+pymysql://root:***@localhost:3306/md_water_services
7 rows affected.


[(89, datetime.datetime(2021, 1, 4, 7, 46), Decimal('137'), 'Monday'),
 (29, datetime.datetime(2021, 1, 5, 7, 58), Decimal('108'), 'Tuesday'),
 (36, datetime.datetime(2021, 1, 6, 6, 21), Decimal('97'), 'Wednesday'),
 (208, datetime.datetime(2021, 1, 7, 6, 46), Decimal('105'), 'Thursday'),
 (15, datetime.datetime(2021, 1, 1, 9, 10), Decimal('120'), 'Friday'),
 (107, datetime.datetime(2021, 1, 2, 7, 7), Decimal('246'), 'Saturday'),
 (103, datetime.datetime(2021, 1, 3, 9, 41), Decimal('82'), 'Sunday')]

In [125]:
%%sql

-- Communicate the information efficiently

SELECT 
    HOUR(time_of_record) AS hour_of_day,
    COUNT(*) AS number_of_visits
FROM visits
GROUP BY hour_of_day
ORDER BY hour_of_day;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
14 rows affected.


[(6, 1528),
 (7, 1592),
 (8, 1620),
 (9, 6555),
 (10, 6502),
 (11, 6517),
 (12, 6463),
 (13, 6648),
 (14, 6576),
 (15, 6420),
 (16, 6494),
 (17, 1574),
 (18, 1620),
 (19, 37)]

In [126]:
%%sql

SELECT 
    TIME_FORMAT(TIME(time_of_record), '%H:00') AS hour_of_day,
    COUNT(*) AS number_of_visits
FROM visits
WHERE time_in_queue <> 0
GROUP BY hour_of_day
ORDER BY hour_of_day;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
14 rows affected.


[('06:00', 1528),
 ('07:00', 1592),
 ('08:00', 1620),
 ('09:00', 6555),
 ('10:00', 6502),
 ('11:00', 6517),
 ('12:00', 6463),
 ('13:00', 6648),
 ('14:00', 6576),
 ('15:00', 6420),
 ('16:00', 6494),
 ('17:00', 1574),
 ('18:00', 1620),
 ('19:00', 37)]

In [130]:
%%sql

SELECT
    TIME_FORMAT(TIME(time_of_record), '%H:00') AS hour_of_day,
    DAYNAME(time_of_record),
    CASE
        WHEN DAYNAME(time_of_record) = 'Sunday' THEN time_in_queue
        ELSE NULL
    END AS Sunday
FROM visits
WHERE
    time_in_queue != 0 -- this exludes other sources with 0 queue times
LIMIT 15;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
15 rows affected.


[('09:00', 'Friday', None),
 ('09:00', 'Friday', None),
 ('10:00', 'Friday', None),
 ('10:00', 'Friday', None),
 ('11:00', 'Friday', None),
 ('11:00', 'Friday', None),
 ('12:00', 'Friday', None),
 ('12:00', 'Friday', None),
 ('13:00', 'Friday', None),
 ('13:00', 'Friday', None),
 ('13:00', 'Friday', None),
 ('14:00', 'Friday', None),
 ('14:00', 'Friday', None),
 ('15:00', 'Friday', None),
 ('16:00', 'Friday', None)]

In [134]:
%%sql

SELECT
    TIME_FORMAT(TIME(time_of_record), '%H:00') AS hour_of_day,
    ROUND(
        AVG(
            CASE
                WHEN DAYNAME(time_of_record) = 'Sunday' THEN NULLIF(time_in_queue, 0)
                ELSE NULL
            END
        ), 
    0) AS Sunday
FROM visits
WHERE time_in_queue != 0
GROUP BY hour_of_day
ORDER BY hour_of_day;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
14 rows affected.


[('06:00', Decimal('79')),
 ('07:00', Decimal('82')),
 ('08:00', Decimal('86')),
 ('09:00', Decimal('84')),
 ('10:00', Decimal('83')),
 ('11:00', Decimal('78')),
 ('12:00', Decimal('78')),
 ('13:00', Decimal('81')),
 ('14:00', Decimal('83')),
 ('15:00', Decimal('83')),
 ('16:00', Decimal('83')),
 ('17:00', Decimal('79')),
 ('18:00', Decimal('80')),
 ('19:00', Decimal('127'))]

In [135]:
%%sql

SELECT
    TIME_FORMAT(TIME(time_of_record), '%H:00') AS hour_of_day,
    ROUND(AVG(CASE WHEN DAYNAME(time_of_record) = 'Sunday' THEN NULLIF(time_in_queue, 0) END), 0) AS Sunday,
    ROUND(AVG(CASE WHEN DAYNAME(time_of_record) = 'Monday' THEN NULLIF(time_in_queue, 0) END), 0) AS Monday,
    ROUND(AVG(CASE WHEN DAYNAME(time_of_record) = 'Tuesday' THEN NULLIF(time_in_queue, 0) END), 0) AS Tuesday,
    ROUND(AVG(CASE WHEN DAYNAME(time_of_record) = 'Wednesday' THEN NULLIF(time_in_queue, 0) END), 0) AS Wednesday,
    ROUND(AVG(CASE WHEN DAYNAME(time_of_record) = 'Thursday' THEN NULLIF(time_in_queue, 0) END), 0) AS Thursday,
    ROUND(AVG(CASE WHEN DAYNAME(time_of_record) = 'Friday' THEN NULLIF(time_in_queue, 0) END), 0) AS Friday,
    ROUND(AVG(CASE WHEN DAYNAME(time_of_record) = 'Saturday' THEN NULLIF(time_in_queue, 0) END), 0) AS Saturday
FROM visits
WHERE time_in_queue != 0
GROUP BY hour_of_day
ORDER BY hour_of_day;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
14 rows affected.


[('06:00', Decimal('79'), Decimal('190'), Decimal('134'), Decimal('112'), Decimal('134'), Decimal('153'), Decimal('247')),
 ('07:00', Decimal('82'), Decimal('186'), Decimal('128'), Decimal('111'), Decimal('139'), Decimal('156'), Decimal('247')),
 ('08:00', Decimal('86'), Decimal('183'), Decimal('130'), Decimal('119'), Decimal('129'), Decimal('153'), Decimal('247')),
 ('09:00', Decimal('84'), Decimal('127'), Decimal('105'), Decimal('94'), Decimal('99'), Decimal('107'), Decimal('252')),
 ('10:00', Decimal('83'), Decimal('119'), Decimal('99'), Decimal('89'), Decimal('95'), Decimal('112'), Decimal('259')),
 ('11:00', Decimal('78'), Decimal('115'), Decimal('102'), Decimal('86'), Decimal('99'), Decimal('104'), Decimal('236')),
 ('12:00', Decimal('78'), Decimal('115'), Decimal('97'), Decimal('88'), Decimal('96'), Decimal('109'), Decimal('239')),
 ('13:00', Decimal('81'), Decimal('122'), Decimal('97'), Decimal('98'), Decimal('101'), Decimal('115'), Decimal('242')),
 ('14:00', Decimal('83'), Decimal('127'), Decimal('104'), Decimal('92'), Decimal('96'), Decimal('110'), Decimal('244')),
 ('15:00', Decimal('83'), Decimal('126'), Decimal('104'), Decimal('88'), Decimal('92'), Decimal('110'), Decimal('248')),
 ('16:00', Decimal('83'), Decimal('127'), Decimal('99'), Decimal('90'), Decimal('99'), Decimal('109'), Decimal('251')),
 ('17:00', Decimal('79'), Decimal('181'), Decimal('135'), Decimal('121'), Decimal('129'), Decimal('151'), Decimal('251')),
 ('18:00', Decimal('80'), Decimal('174'), Decimal('122'), Decimal('113'), Decimal('132'), Decimal('158'), Decimal('240')),
 ('19:00', Decimal('127'), Decimal('159'), Decimal('145'), Decimal('176'), Decimal('137'), Decimal('103'), Decimal('282'))]

In [144]:
%%sql

SELECT
    TIME_FORMAT(TIME(time_of_record), '%H:00') AS hour_of_day,
    ROUND(AVG(CASE WHEN DAYNAME(time_of_record) = 'Sunday' THEN time_in_queue ELSE NULL END), 0) AS Sunday,
    ROUND(AVG(CASE WHEN DAYNAME(time_of_record) = 'Tuesday' THEN time_in_queue ELSE NULL END), 0) AS Tuesday,
    ROUND(AVG(CASE WHEN DAYNAME(time_of_record) = 'Saturday' THEN time_in_queue ELSE NULL END), 0) AS Saturday
FROM
    visits
WHERE
    time_in_queue != 0
GROUP BY
    hour_of_day
ORDER BY
    hour_of_day;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
14 rows affected.


[('06:00', Decimal('79'), Decimal('134'), Decimal('247')),
 ('07:00', Decimal('82'), Decimal('128'), Decimal('247')),
 ('08:00', Decimal('86'), Decimal('130'), Decimal('247')),
 ('09:00', Decimal('84'), Decimal('105'), Decimal('252')),
 ('10:00', Decimal('83'), Decimal('99'), Decimal('259')),
 ('11:00', Decimal('78'), Decimal('102'), Decimal('236')),
 ('12:00', Decimal('78'), Decimal('97'), Decimal('239')),
 ('13:00', Decimal('81'), Decimal('97'), Decimal('242')),
 ('14:00', Decimal('83'), Decimal('104'), Decimal('244')),
 ('15:00', Decimal('83'), Decimal('104'), Decimal('248')),
 ('16:00', Decimal('83'), Decimal('99'), Decimal('251')),
 ('17:00', Decimal('79'), Decimal('135'), Decimal('251')),
 ('18:00', Decimal('80'), Decimal('122'), Decimal('240')),
 ('19:00', Decimal('127'), Decimal('145'), Decimal('282'))]